File: IOS:iPadOS/SnipeMobile/SnipeITQRLink.swift
import Foundation

enum SnipeITQRLink: Equatable {
    case hardware(id: Int)
    case component(id: Int)
    case accessory(id: Int)
    case license(id: Int)
    case consumable(id: Int)
    case location(id: Int)
    case user(id: Int)
    case maintenance(id: Int)
    case hardwareByTag(String)

    /// Parse Snipe-IT item URLs.
    static func parse(_ raw: String) -> SnipeITQRLink? {
        let text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { return nil }

        let query = queryString(from: text)
        let path = pathLike(from: text)
        return parse(path: path, query: query)
    }

    private static let reservedTokens: Set<String> = [
        "bytag", "create", "bulkedit", "bulkdelete", "bulkaudit", "labels",
        "audit", "requested", "clone", "restore", "import", "export",
        "quickscan", "quickadd", "checkout", "checkin", "edit", "delete",
        "view", "files", "history", "maintenances", "api", "v1",
    ]

    private static func parse(path: String, query: String?) -> SnipeITQRLink? {
        if path.localizedCaseInsensitiveContains("hardware"),
           let tag = assetTag(fromQuery: query) {
            return .hardwareByTag(tag)
        }

        let segments = path
            .split(whereSeparator: { $0 == "/" || $0 == "#" })
            .map { decode($0) }
            .filter { !$0.isEmpty }

        for index in 0..<segments.count {
            let segment = segments[index].lowercased()
            let next = index + 1 < segments.count ? segments[index + 1] : nil

            if segment == "hardware", next?.lowercased() == "bytag" {
                let tag = index + 2 < segments.count ? segments[index + 2] : nil
                if let tag, !tag.isEmpty, !isReserved(tag) {
                    return .hardwareByTag(tag)
                }
                if let queryTag = assetTag(fromQuery: query) {
                    return .hardwareByTag(queryTag)
                }
                return nil
            }

            guard let next, !next.isEmpty, !isReserved(next) else { continue }

            if segment == "ht" {
                return .hardwareByTag(next)
            }

            switch segment {
            case "hardware", "assets", "asset":
                return hardwareToken(next)
            case "components", "component":
                if let id = entityId(next) { return .component(id: id) }
            case "accessories", "accessory":
                if let id = entityId(next) { return .accessory(id: id) }
            case "licenses", "license":
                if let id = entityId(next) { return .license(id: id) }
            case "consumables", "consumable":
                if let id = entityId(next) { return .consumable(id: id) }
            case "locations", "location":
                if let id = entityId(next) { return .location(id: id) }
            case "users", "user":
                if let id = entityId(next) { return .user(id: id) }
            case "maintenances", "maintenance":
                if let id = entityId(next) { return .maintenance(id: id) }
            default:
                continue
            }
        }
        return nil
    }

    /// Digits = id; anything else = asset tag.
    private static func hardwareToken(_ token: String) -> SnipeITQRLink {
        if let id = entityId(token) {
            return .hardware(id: id)
        }
        return .hardwareByTag(token)
    }

    private static func entityId(_ token: String) -> Int? {
        guard let id = Int(token), String(id) == token else { return nil }
        return id
    }

    private static func isReserved(_ token: String) -> Bool {
        reservedTokens.contains(token.lowercased())
    }

    private static func pathLike(from text: String) -> String {
        var value = text
        if let scheme = value.range(of: "://") {
            value = String(value[scheme.upperBound...])
            if let hostEnd = value.firstIndex(where: { $0 == "/" || $0 == "?" || $0 == "#" }) {
                value = String(value[hostEnd...])
            } else {
                value = ""
            }
        }
        if let queryStart = value.firstIndex(of: "?") {
            value = String(value[..<queryStart])
        }
        return value.replacingOccurrences(of: "#", with: "/")
    }

    private static func queryString(from text: String) -> String? {
        guard let start = text.firstIndex(of: "?") else { return nil }
        var query = String(text[text.index(after: start)...])
        if let hash = query.firstIndex(of: "#") {
            query = String(query[..<hash])
        }
        return query
    }

    private static func assetTag(fromQuery query: String?) -> String? {
        guard let query, !query.isEmpty else { return nil }
        for part in query.split(separator: "&") {
            let pair = part.split(separator: "=", maxSplits: 1).map(String.init)
            guard pair.count == 2 else { continue }
            let name = pair[0].lowercased()
            guard name == "assettag" || name == "asset_tag" else { continue }
            let tag = decode(Substring(pair[1])).trimmingCharacters(in: .whitespacesAndNewlines)
            if !tag.isEmpty { return tag }
        }
        return nil
    }

    private static func decode(_ value: Substring) -> String {
        let raw = String(value)
        return raw.removingPercentEncoding ?? raw
    }

    func notFoundMessage(id: Int) -> String {
        switch self {
        case .hardware:
            return L10n.string("asset_not_found_id", String(id))
        case .component:
            return L10n.string("component_not_found_id", String(id))
        case .accessory:
            return L10n.string("accessory_not_found_id", String(id))
        case .license:
            return L10n.string("license_not_found_id", String(id))
        case .consumable:
            return L10n.string("consumable_not_found_id", String(id))
        case .location:
            return L10n.string("location_not_found_id", String(id))
        case .user:
            return L10n.string("user_not_found_id", String(id))
        case .maintenance:
            return L10n.string("maintenance_not_found_id", String(id))
        case .hardwareByTag:
            return L10n.string("asset_not_found_id", String(id))
        }
    }
}

File: Android/app/src/main/java/com/callandt/snipemobile/data/api/SnipeITQRLink.kt
package com.callandt.snipemobile.data.api

import com.callandt.snipemobile.ui.util.L10n
import java.net.URLDecoder

sealed class SnipeITQRLink {
    data class Hardware(val id: Int) : SnipeITQRLink()
    data class Component(val id: Int) : SnipeITQRLink()
    data class Accessory(val id: Int) : SnipeITQRLink()
    data class License(val id: Int) : SnipeITQRLink()
    data class Consumable(val id: Int) : SnipeITQRLink()
    data class Location(val id: Int) : SnipeITQRLink()
    data class User(val id: Int) : SnipeITQRLink()
    data class Maintenance(val id: Int) : SnipeITQRLink()
    data class HardwareByTag(val tag: String) : SnipeITQRLink()

    fun notFoundMessage(id: Int): String = when (this) {
        is Hardware, is HardwareByTag -> L10n.string("asset_not_found_id", id.toString())
        is Component -> L10n.string("component_not_found_id", id.toString())
        is Accessory -> L10n.string("accessory_not_found_id", id.toString())
        is License -> L10n.string("license_not_found_id", id.toString())
        is Consumable -> L10n.string("consumable_not_found_id", id.toString())
        is Location -> L10n.string("location_not_found_id", id.toString())
        is User -> L10n.string("user_not_found_id", id.toString())
        is Maintenance -> L10n.string("maintenance_not_found_id", id.toString())
    }

    companion object {
        private val reservedTokens = setOf(
            "bytag", "create", "bulkedit", "bulkdelete", "bulkaudit", "labels",
            "audit", "requested", "clone", "restore", "import", "export",
            "quickscan", "quickadd", "checkout", "checkin", "edit", "delete",
            "view", "files", "history", "maintenances", "api", "v1",
        )

        /** Parse Snipe-IT item URLs. */
        fun parse(urlString: String): SnipeITQRLink? {
            val text = urlString.trim()
            if (text.isEmpty()) return null
            return parsePath(pathLike(text), queryString(text))
        }

        private fun parsePath(path: String, query: String?): SnipeITQRLink? {
            if (path.contains("hardware", ignoreCase = true)) {
                assetTagFromQuery(query)?.let { return HardwareByTag(it) }
            }

            val segments = path
                .split('/', '#')
                .map { decode(it) }
                .filter { it.isNotEmpty() }

            for (index in segments.indices) {
                val segment = segments[index].lowercase()
                val next = segments.getOrNull(index + 1)

                if (segment == "hardware" && next?.equals("bytag", ignoreCase = true) == true) {
                    val tag = segments.getOrNull(index + 2)
                    if (!tag.isNullOrEmpty() && !isReserved(tag)) return HardwareByTag(tag)
                    return assetTagFromQuery(query)?.let { HardwareByTag(it) }
                }

                if (next.isNullOrEmpty() || isReserved(next)) continue

                if (segment == "ht") return HardwareByTag(next)

                when (segment) {
                    "hardware", "assets", "asset" -> return hardwareToken(next)
                    "components", "component" -> entityId(next)?.let { return Component(it) }
                    "accessories", "accessory" -> entityId(next)?.let { return Accessory(it) }
                    "licenses", "license" -> entityId(next)?.let { return License(it) }
                    "consumables", "consumable" -> entityId(next)?.let { return Consumable(it) }
                    "locations", "location" -> entityId(next)?.let { return Location(it) }
                    "users", "user" -> entityId(next)?.let { return User(it) }
                    "maintenances", "maintenance" -> entityId(next)?.let { return Maintenance(it) }
                }
            }
            return null
        }

        /** Digits = id; anything else = asset tag. */
        private fun hardwareToken(token: String): SnipeITQRLink {
            val id = entityId(token)
            return if (id != null) Hardware(id) else HardwareByTag(token)
        }

        private fun entityId(token: String): Int? {
            val id = token.toIntOrNull()
            return if (id != null && id.toString() == token) id else null
        }

        private fun isReserved(token: String): Boolean =
            token.lowercase() in reservedTokens

        private fun pathLike(text: String): String {
            var value = text
            val scheme = value.indexOf("://")
            if (scheme >= 0) {
                value = value.substring(scheme + 3)
                val hostEnd = value.indexOfFirst { it == '/' || it == '?' || it == '#' }
                value = if (hostEnd >= 0) value.substring(hostEnd) else ""
            }
            val queryStart = value.indexOf('?')
            if (queryStart >= 0) value = value.substring(0, queryStart)
            return value.replace('#', '/')
        }

        private fun queryString(text: String): String? {
            val start = text.indexOf('?')
            if (start < 0) return null
            var query = text.substring(start + 1)
            val hash = query.indexOf('#')
            if (hash >= 0) query = query.substring(0, hash)
            return query
        }

        private fun assetTagFromQuery(query: String?): String? {
            if (query.isNullOrEmpty()) return null
            for (part in query.split('&')) {
                val idx = part.indexOf('=')
                if (idx <= 0) continue
                val name = part.substring(0, idx).lowercase()
                if (name != "assettag" && name != "asset_tag") continue
                val tag = decode(part.substring(idx + 1)).trim()
                if (tag.isNotEmpty()) return tag
            }
            return null
        }

        private fun decode(value: String): String =
            runCatching { URLDecoder.decode(value, Charsets.UTF_8.name()) }.getOrDefault(value)
    }
}
